components/selectors/matching.rs

```diff
@@ -541,8 +532,82 @@ where
     }
 }
 
-/// Determines whether the given element matches the given single selector.
 #[inline]
+fn matches_local_name<E>(
+    element: &E,
+    local_name: &LocalName<E::Impl>
+) -> bool
+where
+    E: Element,
+{
+    let name = select_name(
+        element.is_html_element_in_html_document(),
+        &local_name.name,
+        &local_name.lower_name
+    ).borrow();
+    element.get_local_name() == name
+}
+
+/// Determines whether the given element matches the given compound selector.
+#[inline]
+fn matches_compound_selector<E, F>(
+    selector_iter: &mut SelectorIter<E::Impl>,
+    element: &E,
+    context: &mut MatchingContext<E::Impl>,
+    visited_handling: VisitedHandlingMode,
+    flags_setter: &mut F,
+    rightmost: Rightmost,
+) -> bool
+where
+    E: Element,
+    F: FnMut(&E, ElementSelectorFlags),
+{
+    let matches_hover_and_active_quirk =
+        matches_hover_and_active_quirk(&selector_iter, context, rightmost);
+
+    // Handle some common cases first.
```

Comment:
It's a bit unfortunate that we need to do this indeed.

Mind leaving a comment like:

```rust
// TODO(emilio): It'd be nice to just get rid of this, if we make Talos happy / the generic case fast enough.
```

I've seen similar wins today just from enabling stylo-chrome and such (https://treeherder.mozilla.org/perf.html#/compare?originalProject=try&originalRevision=aea8893308877a9139af703db9d5abcc822dddd5&newProject=try&newRevision=e4667f74741b200ae5cc687ce7fb37697b7f394f&framework=1&showOnlyImportant=1). I'm not sure how you feel about landing this vs. waiting. Both look fine to me.

Reply:
The big match needs dispatch anyway, and that may be less friendly to all the predicator and jump cache, so I think unrolling some common cases into a sequential if-chain is a reasonable optimization before rust compiler can do pto.

Reply:
And, it seems to me that commit is just from #19774 plus an `#[inline(never)]` on the helper function. I don't see how that's related to stylo-chrome.